crates/cargo-contract/src/cmd/verify.rs

```diff
@@ -0,0 +1,191 @@
+        let handler = crate::wasm_opt::WasmOptHandler::new(
+            build_info.wasm_opt_settings.optimization_passes,
+            keep_debug_symbols,
+        )?;
+        let wasm_opt_version = handler.version();
+
+        let wasm_opt_matches = wasm_opt_version == expected_wasm_opt_version;
+        let mismatched_wasm_opt = format!(
+            "\nYou are trying to `verify` a contract using `wasm-opt` version `{wasm_opt_version}`.\n\
+             However, the original contract was built using `wasm-opt` version `{expected_wasm_opt_version}`.\n\
+             Please install the matching version and re-run the `verify` command.",
+        );
+        anyhow::ensure!(wasm_opt_matches, mismatched_wasm_opt.bright_yellow());
+
+        let args = ExecuteArgs {
+            manifest_path: manifest_path.clone(),
+            verbosity,
+            build_mode: build_info.build_mode,
+            network: Default::default(),
+            build_artifact: BuildArtifacts::CodeOnly,
+            unstable_flags: Default::default(),
+            optimization_passes: build_info.wasm_opt_settings.optimization_passes,
+            keep_debug_symbols,
+            skip_linting: true,
+            output_type: Default::default(),
+        };
+
+        let build_result = execute(args)?;
+
+        // 3. Grab the built Wasm contract and compare it with the Wasm from the metadata.
+        let reference_wasm = if let Some(wasm) = metadata.source.wasm {
+            wasm
+        } else {
+            anyhow::bail!(
+                "\nThe metadata for the reference contract does not contain a Wasm binary,\n\
+                therefore we are unable to verify the contract."
+                .to_string()
+                .bright_yellow()
+            )
+        };
+
+        let built_wasm_path = if let Some(wasm) = build_result.dest_wasm {
+            wasm
+        } else {
+            // Since we're building the contract ourselves this should always be populated,
+            // but we'll bail out here just in case.
+            anyhow::bail!(
+                "\nThe metadata for the workspace contract does not contain a Wasm binary,\n\
+                therefore we are unable to verify the contract."
+                .to_string()
+                .bright_yellow()
+            )
+        };
+
+        let fs_wasm = std::fs::read(built_wasm_path)?;
+        let built_wasm = SourceWasm::new(fs_wasm);
+
+        if reference_wasm != built_wasm {
+            tracing::debug!(
+                "Expected Wasm Binary '{}'\n\nGot Wasm Binary `{}`",
+                &reference_wasm,
+                &built_wasm
+            );
+
+            anyhow::bail!(format!(
+                "\nFailed to verify the authenticity of {} contract againt the workspace \n\
+                found at {}.",
+                format!("`{}`", metadata.contract.name).bright_white(),
+                format!("{:?}", manifest_path.as_ref()).bright_white()).bright_red()
+            );
```

Comment:
It would be super useful to print out the `SourceCompiler` properties of both `reference_wasm` and `built_wasm` in case of non matching wasm. 

It's useful to know whether they are the same or different for figuring out why verification failed.

And while we are at it, any other properties we have available in `Source` e.g. `SourceLanguage` to catch different versions of `ink!`.